main.go

```diff
@@ -14,191 +14,336 @@ import (
-	Target     string        `json:"target"`
-	Interval   string        `json:"scrape_interval,omitempty"`
-	TimeoutStr string        `json:"timeout,omitempty"`
-	Timeout    time.Duration `json:"-"`
+type ExporterOpts struct {
+	autoDiscover bool
+	interval     time.Duration
+	localAddr    string
+	masterURL    string
 }
 
-func newConfig() (*Config, error) {
-	var config Config
-
-	file, err := ioutil.ReadFile(*configFile)
-	if err != nil {
-		return nil, err
+type PeriodicExporter struct {
+	sync.RWMutex
+	errors  *prometheus.CounterVec
+	metrics []prometheus.Gauge
+	opts    *ExporterOpts
+	slaves  struct {
+		sync.Mutex
+		hostnames []string
 	}
-
-	err = json.Unmarshal(file, &config)
-	return &config, err
 }
 
-func request(e Endpoint) error {
-	client := http.Client{Timeout: 3 * time.Second}
-
-	resp, err := client.Get(e.Target + "/monitor/statistics.json")
-	if err != nil {
-		return err
+func NewMesosExporter(opts *ExporterOpts) *PeriodicExporter {
+	e := &PeriodicExporter{
+		errors: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: "exporter",
+				Name:      "scrape_errors",
+				Help:      "Current total scrape errors ",
+			},
+			[]string{"mesos_slave"},
+		),
+		opts:    opts,
+		metrics: make([]prometheus.Gauge, 0),
 	}
+	e.slaves.hostnames = []string{e.opts.localAddr}
 
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return err
+	if e.opts.autoDiscover {
+		glog.Info("auto discovery enabled from command line flag.")
+
+		// Update nr. of mesos slave every 10 minute
+		e.slaves.hostnames = []string{}
+		go e.runEvery(e.updateSlaves, (10 * time.Minute))
 	}
 
-	var stats []mesos_stats.Monitor
+	// Fetch slave metrics every interval
+	go e.runEvery(e.scrapeSlaves, e.opts.interval)
 
-	err = json.Unmarshal(body, &stats)
-	if err != nil {
-		return err
-	}
+	return e
+}
 
-	for _, t := range stats {
-		cpuLimitVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(t.Statistics.CpusLimit)
-		cpuSysVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(t.Statistics.CpusSystemTimeSecs)
-		cpuUsrVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(t.Statistics.CpusUserTimeSecs)
-		memLimitVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(float64(t.Statistics.MemLimitBytes))
-		memRssVal.WithLabelValues(t.Source, e.Name, t.FrameworkId).Set(float64(t.Statistics.MemRssBytes))
-	}
+func (e *PeriodicExporter) Describe(ch chan<- *prometheus.Desc) {
+	e.rLockMetrics(func() {
+		for _, m := range e.metrics {
+			m.Describe(ch)
+		}
+	})
+	e.errors.MetricVec.Describe(ch)
+}
 
-	return err
+func (e *PeriodicExporter) Collect(ch chan<- prometheus.Metric) {
+	e.rLockMetrics(func() {
+		for _, m := range e.metrics {
+			m.Collect(ch)
+		}
+	})
+	e.errors.MetricVec.Collect(ch)
 }
 
-func requestRunner(e Endpoint) {
-	for {
-		err := request(e)
+func (e *PeriodicExporter) fetch(inCh chan string, outCh chan prometheus.Gauge, wg *sync.WaitGroup) {
```

Comment:
Rather than signaling in/out in the names of the channels, make the direction clear by limiting them to either being read / written by adding `<-` before or after the argument's type in the signature (like `hostChan <-chan string` and `metricsChan chan<- prometheus.Metric`).